### ISSUE TITLE
Discussion - Optimising map backend

### ISSUE THREAD

**Issue by chrismclarke** [OneArmyWorld/onearmy#518]
Firestore is really poorly optimised for these types of operations (costs scale linearly with results retrieved from a query), so getting 1000 small files is really expensive compared to 1 large, but having many rows of data makes a lot more sense for the map. Need to decide best approach to ensure map data is kept fresh but without downloading lots of unnecessary data or incurring high costs.

**Comment by chrismclarke** [comment]
So here's the current system I'm thinking of... It's not great because it has a number of moving parts but hopefully just a temp measure until alternate db option provided. 

- Use 2 databases, _firestore_ for pulling individual records and _realtime-db_ for pulling large batch
- Use local storage caching (manually implemented)
- Use cloud function to keep cloud dbs sync'd
- Keep pin details independent of rest of doc, and retrieve on demand

**Case - User visits page for first time**
Lots of data to retrieve.  Querying all with firestore is costly so instead make a query for all from realtime (cheap). Cache using local storage to retrieve in future

**Case - User revisits map**
Load data from cache, identify newest record and query firestore for anything newer. Commit updates to local storage

**Case - User clicks map pin**
Pull full record from firestore for display

**Case - User adds a bin**
Commit to firestore, use cloud trigger to add copy of data to realtime

If anyone has any thoughts feel free to add.

**Comment by BenGamma** [comment]
> Use 2 databases, firestore for pulling individual records and realtime-db for pulling large batch

Realtime-db is part of firebase also ? Or it means running our own server ?
> Case - User clicks map pin

You mean on click on the marker to open the Card or on click on the Card to go to the user profile ? Because on the profile /u/:username page, a query will be made to load all the user informations. I think on the "large batch" query we could load a minimalist version of a profile to display the markers and the Cards (fields `location`, `profileType`, `profilePicture` & `about`)

The rest sounds good to me, really like the caching feature 👍

**Comment by chrismclarke** [comment]
RealtimeDB is part of firebase, so extra confusing I know. RealtimeDB was the first database that firebase used, and particularly focused on keeping large amounts of data in sync across multiple devices, such as chat apps or games. Firestore came in later to focus more on scaling to huge amounts of data which were less time sensitive (i.e. writing a doc at most once per second, but able to store and instantly retrieve millions-billions), and has much better support for nesting sub-documents, querying data and sync'ing offline. 

We don't really need the full capabilities of either of these databases, but what is important for us is the pricing models. Essentially realtimeDB charges for total bandwidth used whilst firestore charges for total docs retrieved. So if we have 20k map pins, the 3-4MB download is cheap with realtime whilst expensive with firestore, but querying and pulling from a few hundred 'how-tos' for example is much better suited  to firestore.

This becomes very much a moot point when thinking about using other databases (e.g. mongo, neo4j, redis etc.)

**Comment by chrismclarke** [comment]
> You mean on click on the marker to open the Card or on click on the Card to go to the user profile

I mean clicking on the card pulls the card data. Basically when the map first loads we want to be able to see and display all the pins, but for this the only information we actually need is the LatLng of the pin, what type/colour it should be, and a reference to the full document, i.e.

```
{  
    _parent/key: chris-m-clarke,
    location:{lat:24.23423, lng:52352},
    pinType:1
    (or similar)
}
````

Currently each data point holds a lot more information because it has the full pin description, which could easily make each pin 10x bigger in size. Considering most people won't click on most pins, this is a bit wasteful on bandwidth.

So the suggestion will be to click on a pin, and use the parent/key property to retrieve the rest of the data from elsewhere

**Comment by chrismclarke** [comment]
Just seen the existing map frontend is already set up to handle things this way so that's awesome 👍

**Comment by chrismclarke** [comment]
Implemented with pr #524 and added section to wiki about additional database source so closing here
